File: src/fleiber/presents/Referential.java
package fleiber.presents;

import com.google.common.collect.ImmutableMap;
import com.google.common.collect.ImmutableSet;
import com.google.common.collect.ImmutableSetMultimap;
import com.google.common.collect.SetMultimap;

import java.util.Optional;

import static fleiber.presents.Referential.Adult.*;
import static fleiber.presents.Referential.Child.*;

public class Referential {

    public interface Person {
        // just a tag
    }

    public enum Adult implements Person {
        DENIS           ("Denis", "Papé"),
        CATHERINE       ("Catherine", "Mamé"),
        FRANCOIS        ("François"),
        JIE             ("Jie"),
        EMMANUEL        ("Emmanuel", "Manu"),
        HELENE          ("Hélène"),
        LOUIS_MARIE     ("Louis-Marie"),
        ANNE_EMMANUEL   ("Anne-Emmanuel", "Manue"),
        BRUNO           ("Bruno"),
        NELLY           ("Nelly");

        public final String name;
        public final String nickName;

        Adult(String name) {
            this(name, null);
        }
        Adult(String name, String nickName) {
            this.name = name;
            this.nickName = nickName;
        }

        public Optional<String> getNickname() {
            return Optional.ofNullable(nickName);
        }
        public String getDisplayName() {
            return nickName != null ? nickName : name;
        }
        public String getInitals() {
            return name.substring(0, 1) + (name.contains("-") ? name.charAt(name.indexOf("-") + 1) : "");
        }

        @Override
        public String toString() {
            return name;
        }
    }

    public enum Child implements Person {    // Not as in "child of" but more "underage", otherwise Adults should also be Children
        AMELIE      ("Amélie"),
        BARNABE     ("Barnabé"),
        CLAUDE      ("Claude"),
        DANAELLE    ("Danaelle"),
        EMMA        ("Emma"),
        JEANNE      ("Jeanne"),
        PIERRE      ("Pierre"),
        TIMOTHEE    ("Timothée"),
        CYPRIEN     ("Cyprien"),
        ETIENNE     ("Étienne"),
        AUGUSTIN    ("Augustin"),
        GREGOIRE    ("Grégoire"),
        JOSEPH      ("Joseph"),
        ANTOINE     ("Antoine");

        public final String name;

        Child(String name) {
            this.name = name;
        }

        @Override
        public String toString() {
            return name;
        }
    }

    public enum Family {   // Very traditional representation of families :p

        PAPE_MAME       (DENIS, CATHERINE),
        FRANCOIS_JIE    (FRANCOIS, JIE, AMELIE, BARNABE, CLAUDE, DANAELLE, EMMA),
        MANU_HELENE     (EMMANUEL, HELENE, JEANNE, PIERRE),
        LM_MANUE        (LOUIS_MARIE, ANNE_EMMANUEL, TIMOTHEE, CYPRIEN, ETIENNE, AUGUSTIN, GREGOIRE, JOSEPH, ANTOINE),
        BRUNO_NELLY     (BRUNO, NELLY);

        public final Adult father;
        public final Adult mother;
        public final ImmutableSet<Child> children;

        Family(Adult father, Adult mother, Child... children) {
            this.father = father;
            this.mother = mother;
            this.children = ImmutableSet.copyOf(children);
        }

        @Override
        public String toString() {
            return father.getInitals() + "&" + mother.getInitals();
        }


        private static final ImmutableMap<Person,Family> PER_PERSON;
        static {
            ImmutableMap.Builder<Person,Family> builder = ImmutableMap.builder();
            for (Family family : Family.values()) {
                builder.put(family.father, family);
                builder.put(family.mother, family);
                family.children.forEach(child -> builder.put(child, family));
            }
            PER_PERSON = builder.build();
        }

        public static Family getFamily(Person person) {
            return PER_PERSON.get(person);
        }
        public static boolean areMarried(Adult adult1, Adult adult2) {
            return getFamily(adult1).equals(getFamily(adult2));
        }
        public static boolean isChildOf(Adult adult, Child child) {
            return getFamily(adult).equals(getFamily(child));
        }
    }

    public static final SetMultimap<Child,Adult> GOD_PARENTS = ImmutableSetMultimap.<Child,Adult>builder()
            .put(AMELIE, HELENE)
            .put(BARNABE, LOUIS_MARIE)
            .put(CLAUDE, BRUNO)
            .putAll(EMMA, EMMANUEL, ANNE_EMMANUEL)
            .put(JEANNE, FRANCOIS)
            .put(CYPRIEN, JIE)
            .put(ETIENNE, HELENE)
            .put(GREGOIRE, EMMANUEL)
            .put(JOSEPH, FRANCOIS)
            .build();
}

File: src/main/kotlin/fleiber/presents/Referential.kt
package fleiber.presents

import fleiber.presents.Adult.*
import fleiber.presents.Child.*


interface Person {

    val firstName: String

    val initials get() = firstName[0].toString() + firstName.indexOf('-').let { if (it == -1) "" else firstName[it + 1] }
}

enum class Adult(override val firstName: String, val nickName: String? = null) : Person {

    DENIS           ("Denis", "Papé"),
    CATHERINE       ("Catherine", "Mamé"),
    FRANCOIS        ("François"),
    JIE             ("Jie"),
    EMMANUEL        ("Emmanuel", "Manu"),
    HELENE          ("Hélène"),
    LOUIS_MARIE     ("Louis-Marie"),
    ANNE_EMMANUEL   ("Anne-Emmanuel", "Manue"),
    BRUNO           ("Bruno"),
    NELLY           ("Nelly");

    val displayName get() = nickName ?: firstName

    override fun toString() = firstName
}

// Not as in "child of" but more "underage", otherwise Adults should also be Children
// Another possibility would be to have everyone in the same enum, but with a "generation" ordinal
enum class Child(override val firstName: String) : Person {

    AMELIE      ("Amélie"),
    BARNABE     ("Barnabé"),
    CLAUDE      ("Claude"),
    DANAELLE    ("Danaelle"),
    EMMA        ("Emma"),
    JEANNE      ("Jeanne"),
    PIERRE      ("Pierre"),
    TIMOTHEE    ("Timothée"),
    CYPRIEN     ("Cyprien"),
    ETIENNE     ("Étienne"),
    AUGUSTIN    ("Augustin"),
    GREGOIRE    ("Grégoire"),
//    JOSEPH      ("Joseph"),
//    ANTOINE     ("Antoine"),
    TERESA      ("Teresa");

    override fun toString() = firstName
}

// Very traditional representation of families :p
enum class Family(val father: Adult, val mother: Adult, vararg _children: Child) {

    PAPE_MAME       (DENIS, CATHERINE),
    FRANCOIS_JIE    (FRANCOIS, JIE,                 AMELIE, BARNABE, CLAUDE, DANAELLE, EMMA),
    MANU_HELENE     (EMMANUEL, HELENE,              JEANNE, PIERRE, TERESA),
    LM_MANUE        (LOUIS_MARIE, ANNE_EMMANUEL,    TIMOTHEE, CYPRIEN, ETIENNE, AUGUSTIN, GREGOIRE/*, JOSEPH, ANTOINE*/),
    BRUNO_NELLY     (BRUNO, NELLY);

    val children = _children.toSet()

    override fun toString() = "${father.initials}&${mother.initials}"


    companion object {
        internal val BY_PERSON: Map<Person, Family> = mutableMapOf<Person, Family>().apply {
            values().forEach { family ->
                this[family.father] = family
                this[family.mother] = family
                family.children.forEach { child -> this[child] = family }
            }
        }
    }
}


val GOD_PARENTS = listOf(
    AMELIE to HELENE,
    BARNABE to LOUIS_MARIE,
    CLAUDE to BRUNO,
    EMMA to EMMANUEL, EMMA to ANNE_EMMANUEL,
    JEANNE to FRANCOIS,
    TERESA to LOUIS_MARIE,
    CYPRIEN to JIE,
    ETIENNE to HELENE,
    GREGOIRE to EMMANUEL,
//    JOSEPH to FRANCOIS,
).groupBy({ it.first }, { it.second })


fun Adult.isMarriedTo(other: Adult) = family === other.family
fun Adult.isParentOf(child: Child) = family === child.family
fun Adult.isGodParentOf(child: Child) = GOD_PARENTS[child]?.let { this in it } ?: false
val Person.family get() = Family.BY_PERSON[this] ?: error("Oh you poor little soul!")
